netconf.c

```diff
@@ -853,8 +864,512 @@ get_full_tree ()
+    }
+}
+
+static void
+cleanup_xpath_tree (GHashTable *node_table, xmlNode *node, int depth, int *max_depth, bool *root_deleted)
+{
+    xmlNode *cur_node = NULL;
+    xmlNode *next_node = NULL;
+
+    for (cur_node = node; cur_node; cur_node = next_node)
+    {
+        next_node = cur_node->next;
+        if (cur_node->type == XML_ELEMENT_NODE) {
+            if (!g_hash_table_lookup (node_table, cur_node))
+            {
+                xmlUnlinkNode (cur_node);
+                xmlFreeNode (cur_node);
+                if (depth == 0 && cur_node == node)
+                    *root_deleted = true;
+
+                continue;
+            }
+        }
+
+        cleanup_xpath_tree (node_table, cur_node->children, depth + 1, max_depth, root_deleted);
+    }
+    if (depth > *max_depth)
+        *max_depth = depth;
+
+    if (depth == 0)
+        cleanup_empty_branches (node, depth, *max_depth, root_deleted);
+}
+
+static sch_node *
+xpath_node_child (xmlNs *ns, sch_node * parent, const char *child)
+{
+    xmlNode *xml = (xmlNode *) parent;
+    xmlNode *n = xml->children;
+
+    while (n)
+    {
+        if (n->type == XML_ELEMENT_NODE && n->name[0] == 'N')
+        {
+            char *name = (char *) xmlGetProp (n, (xmlChar *) "name");
+            if (name && (name[0] == '*' || sch_match_name (name, child)) && sch_ns_match (n, ns))
+            {
+                xmlFree (name);
+                break;
+            }
+            xmlFree (name);
+        }
+        n = n->next;
+    }
+    return n;
+}
+
+static bool
+_xpath_mark_list_nodes (sch_node *schema, sch_node *node, int flags, int depth, GHashTable *node_table)
+{
+    xmlNode *s;
+    char *name;
+    const char *name_node;
+    bool rc = true;
+
+    for (xmlNode *cur_node = node; cur_node; cur_node = cur_node->next)
+    {
+        if (g_hash_table_lookup (node_table, cur_node))
+        {
+            name_node = (const char *) cur_node->name;
+            s = NULL;
+            for (s = (xmlNode *) schema; s; s = s->next)
+            {
+                name = sch_name (s);
+                if (g_strcmp0 (name, name_node) == 0)
+                    break;
+                g_free (name);
+            }
+
+            if (s)
+            {
+                if (sch_is_list (s))
+                {
+                    xmlNode *first_child = cur_node->children;
+                    if (first_child && !g_hash_table_lookup (node_table, first_child))
+                        g_hash_table_insert (node_table, first_child, first_child);
+                }
+                if (cur_node->children && s->children)
+                {
+                    _xpath_mark_list_nodes (s->children, cur_node->children, flags, depth + 1, node_table);
+                }
+                g_free (name);
+            }
+        }
+    }
+
+    return rc;
+}
+
+bool
+xpath_mark_list_nodes (sch_node * xml, int flags, GHashTable *node_table)
+{
+    sch_node *schema;
+    xmlNode *node = (xmlNode *) xml;
```

Comment:
apteyrx-netconf should not be casting sch_node to xmlNode ... this is part of the abstraction of the schema by apteryx-xml